File: 2022/day9/solve.py
data = [line.strip().split(' ') for line in open('input.txt', 'r')]

def move_head(head, direction):
    match direction:
        case 'U':
            head[1] += 1
        case 'D':
            head[1] -= 1
        case 'L':
            head[0] -= 1
        case 'R':
            head[0] += 1
    
def is_touching(head, tail):
    return abs(head[0] - tail[0]) < 2 and abs(head[1] - tail[1]) < 2

def move_tail(tail, head):
    if head[0] > tail[0]:
        tail[0] += 1
    if head[0] < tail[0]:
        tail[0] -= 1
    if head[1] > tail[1]:
        tail[1] += 1
    if head[1] < tail[1]:
        tail[1] -= 1
    
knots = [[0,0] for _ in range(10)]
visited_spots = set()
for line in data:
    direction = line[0]
    amount = int(line[1])
    for _ in range(amount):
        move_head(knots[0], direction)
        for i in range(1, len(knots)):
            if not is_touching(knots[i], knots[i-1]):
                move_tail(knots[i], knots[i-1])
        visited_spots.add(tuple(knots[9]))

print(len(visited_spots))

File: 2023/day18/solve.go
package main

import (
	"fmt"
	"log"
	"math"
	"saneri/aoc/utils"
	"strconv"
	"strings"
)

type Border struct {
	x, y int
}

func getNextBorder(current Border, direction string, amount int) Border {
	switch direction {
	case "U":
		return Border{current.x, current.y + amount}
	case "R":
		return Border{current.x + amount, current.y}
	case "D":
		return Border{current.x, current.y - amount}
	case "L":
		return Border{current.x - amount, current.y}
	}
	log.Fatal("Invalid direction:", direction)
	return Border{0, 0}
}

func shoelace(borders []Border) int {
	sum1 := 0
	sum2 := 0
	for i := range borders[:len(borders)-1] {
		sum1 += borders[i].x * borders[i+1].y
		sum2 += borders[i].y * borders[i+1].x
	}
	return int(math.Abs(float64(sum1-sum2))) / 2
}

func main() {
	data := utils.ReadInput("input.txt")
	current := Border{0, 0}
	borders := []Border{current}
	borderLength := 0
	for _, line := range data {
		split := strings.Split(line, " ")
		direction := split[0]
		amount, _ := strconv.Atoi(split[1])
		borderLength += amount
		current = getNextBorder(current, direction, amount)
		borders = append(borders, current)
	}
	fmt.Println("a:", shoelace(borders)+borderLength/2+1)

	current = Border{0, 0}
	borders = []Border{current}
	borderLength = 0
	dir := map[string]string{
		"0": "U",
		"1": "R",
		"2": "D",
		"3": "L",
	}
	for _, line := range data {
		split := strings.Split(line, " ")[2]
		split = split[2 : len(split)-1]
		direction := string(split[len(split)-1])
		amount, error := strconv.ParseInt(split[:len(split)-1], 16, 64)
		if error != nil {
			log.Fatal(error)
		}
		borderLength += int(amount)
		current = getNextBorder(current, dir[direction], int(amount))
		borders = append(borders, current)
	}
	fmt.Println("b:", shoelace(borders)+borderLength/2+1)
}
